storage/src/object_store/ecs_s3_client.hpp

```diff
@@ -0,0 +1,412 @@
+   * @brief Initializing underlying libs3. The S3_initialize function must be
+   * called exactly once and only from 1 thread
+   *
+   * @param readOnly
+   */
+  virtual void init(bool readOnly) override {
+    std::lock_guard<std::mutex> g(initLock_);
+    assert(!init_);  // we may return here instead of firing assert failure, but
+                     // probably we want to catch bugs?
+    context_.hostName = config_.url.c_str();
+    context_.protocol = (config_.protocol == "HTTP" ? S3ProtocolHTTP : S3ProtocolHTTPS);
+    context_.uriStyle = S3UriStylePath;
+    /* In ECS terms, this is your object user */
+    context_.accessKeyId = config_.accessKey.c_str();
+    context_.secretAccessKey = config_.secretKey.c_str();
+    /* The name of a bucket to use */
+    context_.bucketName = config_.bucketName.c_str();
+
+    S3Status st = S3_initialize(NULL, S3_INIT_ALL, NULL);
+    if (S3Status::S3StatusOK != st) {
+      LOG_ERROR(logger_, "libs3 init failed, status: " + to_string(st));
+      throw std::runtime_error("libs3 init failed, status: " + to_string(st));
+    }
+    LOG_INFO(logger_, "libs3 initialized");
+    init_ = true;
+  }
+
+  /**
+   * @brief GEt object from the store.
+   *
+   * @param _key key to be retrieved
+   * @param _outValue returned object
+   * @return OK if success, otherwise GeneralError with underlying error status
+   */
+  virtual Status get(const Sliver& _key, OUT Sliver& _outValue) const override {
+    using namespace std::placeholders;
+    Status res = Status::OK();
+    std::function<Status(const Sliver&, OUT Sliver&)> f = std::bind(&EcsS3ClientImpl::get_internal, this, _1, _2);
+    do_with_retry(res, f, _key, _outValue);
+    return res;
+  }
+
+  virtual Status get(const Sliver& _key, OUT char*& buf, uint32_t bufSize, OUT uint32_t& _size) const override {
+    throw std::logic_error("Not implemented for ECS S3 object store");
+  }
+
+  /**
+   * @brief Put object to the store.
+   *
+   * @param _key object's key
+   * @param _value object
+   * @return OK if success, otherwise GeneralError with underlying error status
+   */
+  virtual Status put(const Sliver& _key, const Sliver& _value) override {
+    using namespace std::placeholders;
+    std::function<Status(const Sliver&, const Sliver&)> f = std::bind(&EcsS3ClientImpl::put_internal, this, _1, _2);
+    Status res = Status::OK();
+    do_with_retry(res, f, _key, _value);
+    return res;
+  }
+
+  Status test_bucket() {
+    using namespace std::placeholders;
+    std::function<Status()> f = std::bind(&EcsS3ClientImpl::test_bucket_internal, this);
+    Status res = Status::OK();
+
+    do_with_retry(res, f);
+    return res;
+  }
+
+  Status object_exists(const Sliver& key) {
+    using namespace std::placeholders;
+    std::function<Status(const Sliver&)> f = std::bind(&EcsS3ClientImpl::object_exists_internal, this, _1);
+    Status res = Status::OK();
+    do_with_retry(res, f, key);
+    return res;
+  }
+
+  void wait_for_storage() {
```

Comment:
Just a note for the architectural structure of the rest of the feature: The caller of this function should live in a separate thread from the main thread (replica thread). Otherwise we won't be able to get metrics out of that thread or inspect it in any way, since this will block in endless retry.